src/features/device-state/routes/fleet-state-get-v3.ts

```diff
@@ -0,0 +1,372 @@
+import type { RequestHandler } from 'express';
+import type { Request } from 'express';
+
+import * as _ from 'lodash';
+import {
+	captureException,
+	handleHttpErrors,
+} from '../../../infra/error-handling';
+import {
+	ConfigurationVarsToLabels,
+	EnvVarList,
+	formatImageLocation,
+	readTransaction,
+	rejectUiConfig,
+	setDefaultConfigVariables,
+	varListInsert,
+} from '../state-get-utils';
+import { Expand } from 'pinejs-client-core';
+
+import { sbvrUtils, errors } from '@balena/pinejs';
+import { Application } from '../../../balena-model';
+const { api } = sbvrUtils;
+const { UnauthorizedError } = errors;
+
+type LocalStateApp = FleetStateV3[string]['apps'][string];
+type ServiceComposition = AnyObject;
+export type FleetStateV3 = {
+	[uuid: string]: {
+		name: string;
+		is_managed_by__device?: string;
+		config?: {
+			[varName: string]: string;
+		};
+		apps: {
+			[uuid: string]: {
+				/**
+				 * @deprecated to be removed in state v4
+				 */
+				id: number;
+				name: string;
+				class: 'fleet' | 'block' | 'app';
+				// is_managed_by__device?: string;
+				is_host?: boolean;
+				releases?: {
+					[uuid: string]: {
+						/**
+						 * @deprecated to be removed in state v4
+						 */
+						id: number;
+						services?: {
+							[name: string]: {
+								/**
+								 * @deprecated to be removed in state v4
+								 */
+								id: number;
+								/**
+								 * @deprecated to be removed in state v4
+								 */
+								image_id: number;
+								image: string;
+								/**
+								 * Defaults to true if undefined
+								 */
+								running?: boolean;
+								environment: {
+									[varName: string]: string;
+								};
+								labels: {
+									[labelName: string]: string;
+								};
+								contract?: AnyObject;
+								composition?: ServiceComposition;
+							};
+						};
+						volumes?: AnyObject;
+						networks?: AnyObject;
+					};
+				};
+			};
+		};
+	};
+};
+
+const serviceBuildByImage = (
+	fleet: AnyObject,
+	image?: AnyObject,
+): undefined | AnyObject => {
+	if (image == null) {
+		return;
+	}
+
+	let serviceId: number;
+	if (typeof image.is_a_build_of__service === 'object') {
+		serviceId = image.is_a_build_of__service.__id;
+	} else {
+		serviceId = image.is_a_build_of__service;
+	}
+
+	return fleet.application__has__service_name.find(
+		(fleetService: AnyObject) => fleetService.id === serviceId,
+	);
+};
+
+function buildFleetAppFromRelease(
```

Comment:
I think we should be able to share this entire function with device-state-v3 by just making device an optional parameter there (or I guess not optional but rather `AnyObject | undefined` would make more sense so you have to explicitly pass undefined)

Reply:
yep, applied also for config.. Thx